spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/integration/inbound/PubSubMessageSource.java

```diff
@@ -0,0 +1,145 @@
+			String subscriptionName) {
+		Assert.notNull(pubSubSubscriberOperations, "Pub/Sub subscriber template can't be null.");
+		Assert.notNull(subscriptionName, "Pub/Sub subscription name can't be null.");
+		this.pubSubSubscriberOperations = pubSubSubscriberOperations;
+		this.subscriptionName = subscriptionName;
+	}
+
+	public void setAckMode(AckMode ackMode) {
+		Assert.notNull(ackMode, "The acknowledgement mode can't be null.");
+		this.ackMode = ackMode;
+	}
+
+	public void setPayloadType(Class payloadType) {
+		Assert.notNull(payloadType, "The payload type cannot be null.");
+		this.payloadType = payloadType;
+	}
+
+	public void setHeaderMapper(HeaderMapper<Map<String, String>> headerMapper) {
+		Assert.notNull(headerMapper, "The header mapper can't be null.");
+		this.headerMapper = headerMapper;
+	}
+
+	/**
+	 * Instructs synchronous pull to wait until at least one message is available.
+	 */
+	public void setBlockOnPull() {
+		this.blockOnPull = true;
+	}
+
+	/**
+	 * Provides a single polled message.
+	 * <p>Messages are received from Pub/Sub by synchronous pull, in batches determined
+	 * by {@code fetchSize}.
+	 * @param fetchSize number of messages to fetch from Pub/Sub.
+	 * @return {@link Message} wrapper containing the original message.
+	 */
+	@Override
+	protected Object doReceive(int fetchSize) {
+		if (this.cachedMessages.isEmpty()) {
+			Integer maxMessages = (fetchSize > 0) ? fetchSize : 1;
+
+			List<ConvertedAcknowledgeablePubsubMessage> messages
+					= this.pubSubSubscriberOperations.pullAndConvert(this.subscriptionName, maxMessages, !this.blockOnPull, this.payloadType);
+			if (messages.isEmpty()) {
+				return null;
+			}
+			else if (messages.size() == 1) {
+				// don't bother storing.
+				return processMessage(messages.get(0));
+			}
+			else {
+				this.cachedMessages.addAll(messages);
+			}
+		}
+
+		return processMessage(this.cachedMessages.pollFirst());
+	}
+
+	@Override
+	public String getComponentType() {
+		return "gcp-pubsub:message-source";
+	}
+
+	/**
+	 * Applies header customizations and acknowledges the message, if necessary.
+	 * <p>{@link AckMode#AUTO} and {@link AckMode#AUTO_ACK} result in automatic acking on
+	 * success. {@link AckMode#AUTO} results in automatic nacking on failure.
+	 * @param message source Pub/Sub message.
+	 * @return {@link Message} wrapper containing the original message.
+	 */
+	private Message<Object> processMessage(ConvertedAcknowledgeablePubsubMessage message) {
+
+		Map<String, Object> messageHeaders =
+				this.headerMapper.toHeaders(message.getPubsubMessage().getAttributesMap());
+
+		messageHeaders.put(GcpPubSubHeaders.ORIGINAL_MESSAGE, message);
+		messageHeaders.put(IntegrationMessageHeaderAccessor.ACKNOWLEDGMENT_CALLBACK,
+					new PubSubAcknowledgmentCallback(message, this.ackMode));
+
+		return getMessageBuilderFactory()
+				.withPayload(message.getPayload())
+				.copyHeaders(messageHeaders)
+				.build();
```

Comment:
Consider return an `AbstractIntegrationMessageBuilder` instead.
See `evaluateHeaders()` in the `AbstractMessageSource`. So, with additional header we won't be need to rebuilt into a new `Message`.

In other words: don't call this `build()`.